src/track/beats.cpp

```diff
@@ -53,3 +190,676 @@ double Beats::findNBeatsFromSample(double fromSample, double beats) const {
     return nthBeat;
 };
 
+void Beats::onBeatlistChanged() {
+    if (!isValid()) {
+        m_dLastFrame = 0;
+        m_dCachedBpm = 0;
+        return;
+    }
+    m_dLastFrame = m_beats.last().frame_position();
+    track::io::Beat startBeat = m_beats.first();
+    track::io::Beat stopBeat = m_beats.last();
+    m_dCachedBpm = calculateBpm(startBeat, stopBeat);
+}
+
+bool Beats::isValid() const {
+    return m_iSampleRate > 0 && m_beats.size() > 0;
+}
+
+double Beats::calculateBpm(const track::io::Beat& startBeat, const track::io::Beat& stopBeat) const {
+    if (startBeat.frame_position() > stopBeat.frame_position()) {
+        return -1;
+    }
+
+    BeatList::const_iterator curBeat =
+            std::lower_bound(m_beats.cbegin(), m_beats.cend(), startBeat, BeatLessThan);
+
+    BeatList::const_iterator lastBeat =
+            std::upper_bound(m_beats.cbegin(), m_beats.cend(), stopBeat, BeatLessThan);
+
+    QVector<double> beatvect;
+    for (; curBeat != lastBeat; ++curBeat) {
+        const track::io::Beat& beat = *curBeat;
+        if (beat.enabled()) {
+            beatvect.append(beat.frame_position());
+        }
+    }
+
+    if (beatvect.isEmpty()) {
+        return -1;
+    }
+
+    return BeatUtils::calculateBpm(beatvect, m_iSampleRate, 0, 9999);
+}
+
+double Beats::findPrevBeatNew(FrameNum frame) const {
+    return findNthBeatNew(frame, -1);
+}
+
+bool Beats::findPrevNextBeatsNew(FrameNum frame,
+        FrameNum* pPrevBeatFrame,
+        FrameNum* pNextBeatFrame) const {
+    QMutexLocker locker(&m_mutex);
+
+    if (!isValid()) {
+        *pPrevBeatFrame = -1;
+        *pNextBeatFrame = -1;
+        return false;
+    }
+
+    track::io::Beat beat;
+    beat.set_frame_position(frame);
+
+    // it points at the first occurrence of beat or the next largest beat
+    BeatList::const_iterator it =
+            std::lower_bound(m_beats.cbegin(), m_beats.cend(), beat, BeatLessThan);
+
+    // If the position is within 1/10th of a second of the next or previous
+    // beat, pretend we are on that beat.
+    const double kFrameEpsilon = 0.1 * m_iSampleRate;
```

Comment:
I see many of these literals scattered all over the code. Please extract those constants for documentation! Inline comments are not sufficient.